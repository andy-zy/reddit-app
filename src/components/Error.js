// @flow
import React, { Component } from 'react'
import {
  Text,
  View,
} from 'react-native'

import type { PropsBaseT } from '../domain/types'

import { layout, colors } from '../styles'

export default class Error extends Component<PropsBaseT> {

  static navigationOptions = {
    title: 'Error',
  }

  render() {
    const { navigation } = this.props
    const error: string = navigation.state.params.error || 'Something went wrong!'

    return (
      <View style={layout.vContainer}>
        <Text style={colors.error}>{error}</Text>
      </View>
    )
  }
}