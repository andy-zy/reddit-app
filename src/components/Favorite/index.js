import Favorite from './Favorite.container'

export default Favorite