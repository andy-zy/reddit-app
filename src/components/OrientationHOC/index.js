import OrientationHOC from './OrientationHOC'

export default OrientationHOC