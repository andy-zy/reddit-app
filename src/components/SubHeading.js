// @flow
import React from 'react'
import { Text } from 'react-native'

import { layout } from '../styles'

export default (props: Object) => <Text {...props} style={[layout.subHeading, props.style]}>{props.children}</Text>