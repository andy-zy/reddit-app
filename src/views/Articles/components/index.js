import Tabs from './Tabs'
import Search from './Search'
import ArticleList from './ArticleList'
import ListItem from './ListItem'

export {
  Tabs,
  Search,
  ArticleList,
  ListItem,
}