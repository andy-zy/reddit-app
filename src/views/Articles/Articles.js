// @flow
import React, { Component } from 'react'
import {
  View,
  Image,
  FlatList,
  ActivityIndicator,
  TouchableHighlight,
} from 'react-native'

import type { ArticleT } from '../../domain/types'
import type { PropsT, RenderT } from './types'

import { SubHeading } from '../../components'
import { layout } from '../../styles';
import styles from './styles'

export default class Articles extends Component<PropsT> {

  static navigationOptions = {
    title: 'ReactJS',
  }

  componentDidMount() {
    const { getArticlesByCategory } = this.props

    getArticlesByCategory(Articles.navigationOptions.title.toLocaleLowerCase())
  }

  renderItem = ({ item }: RenderT) => {
    const { orientation } = this.props
    const isPortrait: boolean = orientation === 'PORTRAIT'
    const imageSource = item.preview &&
      item.preview.images &&
      item.preview.images[0] &&
      item.preview.images[0].source;

    return (
      <TouchableHighlight onPress={this.handlePress(item)} style={styles.button}>
        <View style={[styles.item, styles.button, !isPortrait && styles.itemL]}>
          <View style={[styles.imageWrap, isPortrait ? styles.imageWrapP : styles.imageWrapL]} >
            {
              imageSource ? (
                <Image
                  source={{ uri: imageSource.url, height: imageSource.height, width: imageSource.width }}
                  style={styles.image}
                />
              ) : (
                <Image source={require('../../images/No_Image_Available.png')} style={styles.image} />
              )
            }
          </View>
          <SubHeading style={!isPortrait && styles.textL}>{item.title}</SubHeading>
        </View>
      </TouchableHighlight>
    )
  }

  renderSeparator = () => <View style={styles.separator} />

  keyExtractor = (item: ArticleT) => item.id

  handlePress = (item: ArticleT) => {
    const { setActiveArticle, navigation } = this.props

    return () => {
      setActiveArticle(item)
      navigation.navigate('Browser')
    }
  }

  render() {
    const {
      articles,
      isFetching,
      orientation,
    } = this.props

    const isPortrait: boolean = orientation === 'PORTRAIT'

    return (
      <View style={isPortrait ? layout.vContainer : layout.hContainer}>
        {
          isFetching ? <ActivityIndicator /> : (
            <FlatList
              data={articles}
              renderItem={this.renderItem}
              keyExtractor={this.keyExtractor}
              ItemSeparatorComponent={this.renderSeparator}
            />
          )
        }
      </View>
    )
  }
}
