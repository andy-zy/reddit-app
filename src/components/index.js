import { default as Favorite } from './Favorite'
import { default as OrientationHOC } from './OrientationHOC'

export {
  Favorite,
  OrientationHOC,
}