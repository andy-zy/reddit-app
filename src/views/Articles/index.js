import Articles from './Articles.container'

export default Articles