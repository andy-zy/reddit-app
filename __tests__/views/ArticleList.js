import 'react-native'
import React from 'react'
import {
  View,
  Text,
} from 'react-native'
import { shallow } from 'enzyme'
import renderer from 'react-test-renderer'
import { ArticleList } from '../../src/views/Articles/components'

describe('Error', () => {
  let props

  beforeEach(() => {
    props = {
      error: '',
      isFetching: false,
      articles: [
        { id: 'id1', title: 'title1', url: 'url1', preview: {} },
        { id: 'id2', title: 'title2', url: 'url2', preview: {} },
      ],
      itemRenderer: (item) => <View><Text>{item.title}</Text></View>,
      fetchData: jest.fn(),
      onRefresh: jest.fn(),
    }
  })

  it('renders correctly with data', () => {
    const tree = renderer.create(<ArticleList {...props} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders correctly when data is fetching', () => {
    const tree = renderer.create(<ArticleList {...props} isFetching />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('contains <FlatList />', () => {
    const component = shallow(<ArticleList {...props} />)
    expect(component.find('FlatList')).toHaveLength(1)
  })

  it('<FlatList /> refreshing should call onRefresh', () => {
    const component = shallow(<ArticleList {...props} />)
    component.find('FlatList').simulate('refresh')
    expect(props.onRefresh).toHaveBeenCalled()
  })

  it('<FlatList /> reaching end should call fetchData', () => {
    const component = shallow(<ArticleList {...props} />)
    component.find('FlatList').simulate('endReached')
    expect(props.fetchData).toHaveBeenCalled()
  })
})