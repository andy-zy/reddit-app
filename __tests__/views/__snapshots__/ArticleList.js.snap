// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Error renders correctly when data is fetching 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RCTScrollView
    ItemSeparatorComponent={[Function]}
    data={
      Array [
        Object {
          "id": "id1",
          "preview": Object {},
          "title": "title1",
          "url": "url1",
        },
        Object {
          "id": "id2",
          "preview": Object {},
          "title": "title2",
          "url": "url2",
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReached={[MockFunction]}
    onEndReachedThreshold={0.5}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onRefresh={[MockFunction]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    refreshControl={
      <RefreshControlMock
        onRefresh={[MockFunction]}
        progressViewOffset={undefined}
        refreshing={true}
      />
    }
    refreshing={true}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <RCTRefreshControl />
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View>
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
          />
        </View>
        <View
          style={
            Object {
              "height": 10,
            }
          }
        />
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View>
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
          />
        </View>
      </View>
    </View>
  </RCTScrollView>
  <View
    style={
      Object {
        "height": 24,
      }
    }
  >
    <ActivityIndicator
      animating={true}
      color="#999999"
      hidesWhenStopped={true}
      size="small"
    />
  </View>
</View>
`;

exports[`Error renders correctly with data 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RCTScrollView
    ItemSeparatorComponent={[Function]}
    data={
      Array [
        Object {
          "id": "id1",
          "preview": Object {},
          "title": "title1",
          "url": "url1",
        },
        Object {
          "id": "id2",
          "preview": Object {},
          "title": "title2",
          "url": "url2",
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReached={[MockFunction]}
    onEndReachedThreshold={0.5}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onRefresh={[MockFunction]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    refreshControl={
      <RefreshControlMock
        onRefresh={[MockFunction]}
        progressViewOffset={undefined}
        refreshing={false}
      />
    }
    refreshing={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <RCTRefreshControl />
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View>
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
          />
        </View>
        <View
          style={
            Object {
              "height": 10,
            }
          }
        />
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View>
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
          />
        </View>
      </View>
    </View>
  </RCTScrollView>
  <View
    style={
      Object {
        "height": 24,
      }
    }
  />
</View>
`;
